### PR TITLE
refactor(x/sync): remove need for passing proof nodes to completeWorkItem

### PR DESCRIPTION
**Pull request by alarso16**
## Why this should be merged

As a part of the refactor to allow Firewood to use x/sync, we should remove any proof logic from sync manager. This moves finding the next key to request to the proof commit implementation, as well as parsing the proof from bytes.

## How this works

The `completeWorkItem` function does not need to know the largest handled key, but rather the next key that needs retrieved. This is handled through a hook, since even on some non-error returns, it shouldn't be called (see implementation). Instead of passing in the `DB` to the Sync Manager, it will take this `ProofClient`

## How this was tested

Unit tests.

## Need to be documented in RELEASES.md?

No.

### CODE REVIEW

File: x/sync/manager.go [hunk 1]
```diff
@@ -642,14 +656,25 @@ func (m *Manager) handleChangeProofResponse(
...
-		m.completeWorkItem(ctx, work, largestHandledKey, targetRootID, rangeProof.EndProof)
 	default:
 		return fmt.Errorf(
 			"%w: %T",
 			errUnexpectedChangeProofResponse, changeProofResp,
 		)
 	}
+	// Find the next key to fetch.
+	// If this is empty, then we have no more keys to fetch.
+	if !largestHandledKey.IsNothing() {
```
[the remark this answers is not in the document]
Comment (alarso16) [reply x/sync/manager.go]: I will be pushing this down a layer in a later PR

File: x/sync/sync_test.go [hunk 2]
```diff
@@ -422,6 +422,55 @@ func Test_Sync_FindNextKey_ExtraValues(t *testing.T) {
...
+		ChangeProofClient:     p2ptest.NewSelfClient(t, ctx, ids.EmptyNodeID, NewGetChangeProofHandler(db)),
+		TargetRoot:            targetRoot,
+		SimultaneousWorkLimit: 5,
+		Log:                   logging.NoLog{},
+		BranchFactor:          merkledb.BranchFactor16,
+	}, prometheus.NewRegistry())
+	require.NoError(err)
+
+	// Since both keys are identical, the next key should be nothing, since the range is complete
+	nextKey, err := syncer.findNextKey(context.Background(), testKey, maybe.Some([]byte{0x11, 0x11}), proof.EndProof)
```
Comment (joshua-kim) [inline x/sync/sync_test.go]: Why are we calling unexported code to test this? Is it not possible for us to test this w/ the exported api? Testing package implementation details tends to be brittle + not representative of how callers use this package

Comment (alarso16) [reply x/sync/sync_test.go]: This is just ensuring an edge case in the `findNextKey` function is handled correctly. As you point out, if this function is going to be tested thoroughly, all the tests should be redesigned to ensure that these edge cases are hit and handled correctly internally

File: x/merkledb/db.go [hunk 3]
```diff
@@ -1390,3 +1429,217 @@ func cacheEntrySize(key Key, n *node) int {
...
+	// or have the same ID in the other proof.
+	proofKeyPath := ToKey(lastReceivedKey)
+
+	// If the received proof is an exclusion proof, the last node may be for a
+	// key that is after the [lastReceivedKey].
+	// If the last received node's key is after the [lastReceivedKey], it can
+	// be removed to obtain a valid proof for a prefix of the [lastReceivedKey].
+	if !proofKeyPath.HasPrefix(endProof[len(endProof)-1].Key) {
+		endProof = endProof[:len(endProof)-1]
+		if len(endProof) == 0 {
```
Comment (alarso16) [inline x/merkledb/db.go:1473]: This was a missed edge case prior to these tests

File: x/merkledb/generator.go [hunk 4]
```diff
@@ -0,0 +1,86 @@
+// Copyright (C) 2019-2024, Ava Labs, Inc. All rights reserved.
```
Comment (alarso16) [inline x/merkledb/generator.go:1]: Moved from `x/sync/sync_test.go`

File: x/sync/merkledb_client.go [hunk 5]
```diff
@@ -0,0 +1,479 @@
...
+//
+// [rangeEnd] is the end of the range that we want to fetch.
+//
+// Returns Nothing if there are no more keys to fetch in [lastReceivedKey, rangeEnd].
+//
+// [endProof] is the end proof of the last proof received.
+//
+// Invariant: [lastReceivedKey] < [rangeEnd].
+// If [rangeEnd] is Nothing it's considered > [lastReceivedKey].
+func (c *xSyncClient) findNextKey(
```
Comment (alarso16) [inline x/sync/merkledb_client.go]: Everything this function and below was completely unmodified.

File: x/sync/proof_test.go [hunk 6]
```diff
@@ -0,0 +1,657 @@
...
+	"github.com/ava-labs/avalanchego/ids"
+	"github.com/ava-labs/avalanchego/network/p2p/p2ptest"
+	"github.com/ava-labs/avalanchego/utils/logging"
+	"github.com/ava-labs/avalanchego/utils/maybe"
+	"github.com/ava-labs/avalanchego/x/merkledb"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/stretchr/testify/require"
+)
+
+func Test_Sync_FindNextKey_InSync(t *testing.T) {
```
Comment (alarso16) [inline x/sync/proof_test.go:26]: These tests are unmodified besides initialization

File: x/sync/db.go [hunk 7]
```diff
@@ -12,3 +19,18 @@ type DB interface {
 	merkledb.ChangeProofer
 	merkledb.RangeProofer
 }
+
+type ProofClient interface {
```
Comment (alarso16) [inline x/sync/db.go]: Any ideas for a better name?

Comment (rkuris) [reply x/sync/db.go]: SyncClient?

File: x/sync/db.go [hunk 8]
```diff
@@ -12,3 +19,16 @@ type DB interface {
...
+	merkledb.Clearer
+	merkledb.MerkleRootGetter
```
Comment (rkuris) [inline x/sync/db.go]: nit: naming: These are no longer merkledb interfaces, correct? Sure, they are implemented by merkledb but IIRC golang just lets you declare interfaces. If so, interfaces used by ProofClient (or SyncClient) should be maybe in a package named xsync or something.

File: x/sync/manager.go [hunk 9]
```diff
@@ -134,24 +131,20 @@ type Manager struct {
...
-	tokenSize int
 
 	stateSyncNodeIdx uint32
 	metrics          SyncMetrics
 }
 
 // TODO remove non-config values out of this struct
 type ManagerConfig struct {
-	DB                    DB
+	ProofClient           ProofClient
```
Comment (rkuris) [inline x/sync/manager.go]: nit: Does it make sense to just use `ProofClient` here so you aren't stuttering? Not sure if that makes sense.

File: x/sync/manager.go [hunk 10]
```diff
@@ -134,24 +131,20 @@ type Manager struct {
...
 	RangeProofClient      *p2p.Client
 	ChangeProofClient     *p2p.Client
 	SimultaneousWorkLimit int
 	Log                   logging.Logger
 	TargetRoot            ids.ID
-	BranchFactor          merkledb.BranchFactor
```
Comment (rkuris) [inline x/sync/manager.go:151]: What happened to this?

File: x/sync/manager.go [hunk 11]
```diff
@@ -327,7 +315,7 @@ func (m *Manager) requestChangeProof(ctx context.Context, work *workItem) {
 
 	if work.localRootID == targetRootID {
 		// Start root is the same as the end root, so we're done.
-		m.completeWorkItem(ctx, work, work.end, targetRootID, nil)
+		m.completeWorkItem(work, work.end, targetRootID)
```
Comment (rkuris) [inline x/sync/manager.go:319]: 🏆 seems like not having context means we aren't doing any database work down there, which I think is good!

File: x/sync/proof_test.go (hunk 6 above)
Comment (rkuris) [reply x/sync/proof_test.go:26]: These comments significantly reduce review time, ty for doing that!

File: x/sync/db.go (hunk 7 above)
Comment (joshua-kim) [reply x/sync/db.go]: What do we define this interface for?

File: x/sync/merkledb_client.go [hunk 12]
```diff
@@ -0,0 +1,478 @@
...
+)
+
+var _ ProofClient = (*xSyncClient)(nil)
+
+type ClientConfig struct {
+	Hasher       merkledb.Hasher
+	BranchFactor merkledb.BranchFactor
+}
+
+type xSyncClient struct {
```
Comment (joshua-kim) [inline x/sync/merkledb_client.go]: This abstraction feels very difficult to reason about and there doesn't seem to be a clear abstraction boundary defined - what is this type intended to represent and what is it's responsibility? It's not really a client - it doesn't make any requests, but seems to define handlers for how to interpret responses + you can interact with the underlying database it's coupled to.

File: x/sync/db.go (hunk 7 above)
Comment (alarso16) [reply x/sync/db.go]: The x/sync manager should be able to work with other database implementations besides `MerkleDB` (specifically, [Firewood](https://github.com/ava-labs/firewood)). This is the smallest reasonable interface to allow other databases to hook in.

File: x/sync/merkledb_client.go (hunk 12 above)
Comment (alarso16) [reply x/sync/merkledb_client.go]: It isn't really a client, but I didn't know what to call it. The goal was to remove the dependence on database and proof implementation (specifically `merkledb`) from the sync manager. The x/sync manager in particular shouldn't depend on any specific database/proof implementation

File: x/sync/manager.go (hunk 10 above)
Comment (alarso16) [reply x/sync/manager.go:151]: Moved to `merkleDBSyncClient`. It was only used for db ops

File: x/sync/manager.go (hunk 9 above)
Comment (alarso16) [reply x/sync/manager.go]: This actually does something different - it embeds the struct.
Here I made a little [demo](https://go.dev/play/p/2zC1MBa-0rg)

File: x/sync/db.go (hunk 7 above)
Comment (alarso16) [reply x/sync/db.go]: How about `DBSyncClient`?

File: x/sync/db.go [hunk 13]
```diff
@@ -3,12 +3,23 @@
...
+	GetRootHash(ctx context.Context) (ids.ID, error)
+	HandleRangeProofResponse(ctx context.Context, request *pb.SyncGetRangeProofRequest, responseBytes []byte) (maybe.Maybe[[]byte], error)
+	HandleChangeProofResponse(
+		ctx context.Context,
+		request *pb.SyncGetChangeProofRequest,
+		responseBytes []byte,
+	) (maybe.Maybe[[]byte], error)
+	Error() error
+	Clear() error
 }
```
Comment (joshua-kim) [inline x/sync/db.go:25]: Thoughts on this interface are:
1. The abstraction boundary isn't clear, we couple the database w/ how responses are handled
2. `Error` is strange IMO, we're expecting our sync package to just keep checking this return value. We could probably get rid of this from the interface if we moved this information somewhere else... we would only fatal during response handling so does it seem reasonable to see if it's possible to expose that information to that API?

File: x/sync/proof_test.go (hunk 6 above)
Comment (joshua-kim) [reply x/sync/proof_test.go:26]: I would avoid unnecessary moves if possible in this PR, it makes it difficult to review - it's okay (and common) in Go to just stuff things in one file.

**Comment by alarso16** [comment]
Closed in favor of smaller PRs
